sdks/python/apache_beam/coders/coders.py

```diff
@@ -210,11 +216,15 @@ def as_cloud_object(self):
   def __repr__(self):
     return self.__class__.__name__
 
+  # pylint: disable=protected-access
   def __eq__(self, other):
-    # pylint: disable=protected-access
     return (self.__class__ == other.__class__
             and self._dict_without_impl() == other._dict_without_impl())
-    # pylint: enable=protected-access
+
+  def __hash__(self):
```

Comment:
Is this change required by Python3 migration or we are just fixing an omission that __hash__ was not previously defined, while __eq__ was?

Reply:
On Python 2, `__hash__` defaults to give a different value for each instance. On Python 3, `__hash__` defaults to `None` if `__eq__` is implemented. By implementing `__hash__`, we get consistent behavior on both versions.

Reply:
Revisiting this in light of other PRs. I think, it would be safer to guarantee the contract that hash does not change for the same object if we compute it here based on object type, sent https://github.com/apache/beam/pull/5390. 
Another possibility to guarantee consistent behavior between Python 2 and 3 would be to set `__hash__ = None` if we can infer that a class is obviously non-hashable.

Reply:
We can also explicitly implement the default python 2 hash, which is guaranteed to stay the same for an object:
`hash(id(self))`

Reply:
That's true. Although that wouldn't honor the contract between __eq__ and __hash__.